Trigger `tx` catchup when needed

hewison-chris opened the pull request:
In the case where a node receives a `tx set` with at least one unknown `tx`.
Now also includes the signature catchup from PR https://github.com/bosagora/agora/pull/2957.

hewison-chris committed:
Trigger signature catchup when needed

In the case where the previous block has less than majority signatures
but it is time to start nominations for the next block.

hewison-chris committed:
Trigger tx catchup timer when needed

In the case where a node receives a `tx set` with at least one unknown
`tx`.

linked0 commented on source/agora/network/Manager.d:
```diff
@@ -853,6 +854,11 @@ public class NetworkManager
     public void getUnknownTXs (NodeLedger ledger) @safe nothrow
     {
+        // Protect against running multiple times simultaneously
+        if (this.isGettingUnknownTXs)
```
What happened if a race condition occurs? The `getUnknownTXs` is called in two places, so I think it might happen.

hewison-chris replied:
Good point, I'll add `synchronized`.

hewison-chris replied:
Updated with `shared bool` and `synchronized` block for checking and setting the value.

hewison-chris replied:
Reverted as they are called from the same thread.

linked0 reviewed (approved):
LGTM

Geod24 commented on source/agora/test/RestoreSlashingInfo.d:
```diff
@@ -74,16 +74,23 @@ unittest
-    Thread.sleep(3.seconds);
-    // Set B validators should discover each other
-    network.waitForDiscovery();
+    Thread.sleep(conf.node.network_discovery_interval);
```
How's that different from `waitForDiscovery` ?

hewison-chris replied:
The `waitForDiscovery` can return true before the outsiders have registered again after restart.

hewison-chris replied:
^ I do not quite understand why though. So I'll make a PR with just the first commit.

hewison-chris replied:
This commit has been removed

Geod24 reviewed (approved):
Last 2 commits LGTM, would like a more extensive explanation on the first one.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -916,6 +916,7 @@ extern(D):
             if (fail_reason == this.ledger.InvalidConsensusDataReason.NotInPool)
             {
                 log.trace("validateValue(): This node can not yet fully validate this value: {}. Data: {}", fail_reason, data.prettify);
+                this.network.getUnknownTXs(this.ledger);
```
Having a validation method trigger a bunch of network requests and yield is a bomb (impatiently) waiting to go off.

hewison-chris replied:
Updated to set a timer which will start a `Task` which calls the `getUnkownTXs`.

hewison-chris commented:
🟢

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -235,6 +246,12 @@ extern(D):
         }
     }
 
+    /// Trigger network cathup of unknown txs
+    private void txCatchupTask ()
+    {
+        this.network.getUnknownTXs(this.ledger);
+    }
+
```
What I mentioned last time was trying to use this timer:
https://github.com/bosagora/agora/blob/31b9fd794b703e98b8188f57b13ddce242b51b17/source/agora/node/FullNode.d#L460-L461

Otherwise we risk having two timers doing the same thing that trigger at the same time.
You might need to split what the timer do, but we really shouldn't have two timers doing this.

hewison-chris replied:
I can create a new timer for `getUnknownTX` in `FullNode` and pass it to the`Nominator` class.

hewison-chris replied:
Pushed an update with new timer for `tx catchup` and pass it to `Nominator`. I expect I will need to split some of the changes into a separate commit. But want to see if it passes Ci first.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -490,6 +496,7 @@ extern(D):
             this.log.trace(
                 "checkNominate(): Last block ({}) doesn't have majority signatures, signed={}",
                 this.ledger.height(), this.ledger.lastBlock().header.validators);
+            this.network.getMissingBlockSigs(this.ledger, this.acceptHeader);
```
Doesn't this also have the yield problem ?

hewison-chris replied:
But this one is only called when we are blocked because we do not have the signatures for the previous block. So it really is a priority to catch up before anything else.

Geod24 replied:
Alright

hewison-chris committed:
Trivial: Log height for reduced val rewards

hewison-chris committed:
Trigger signature catchup for payout period

If we have a different `Coinbase` hash then most likely we need to
catchup signatures for that payout period. We also clear the cached
Coinbase tx so that it can be re calculated.

hewison-chris commented:
Now included in https://github.com/bosagora/agora/pull/3000
